File: Ahorn/entities/zipline.jl
module IsaGrabBagZipline

using ..Ahorn, Maple

@mapdef Entity "isaBag/zipline" Zipline(x::Integer, y::Integer, nodes::Array{Tuple{Integer, Integer}, 1}=Tuple{Integer, Integer}[], usesStamina::Bool=true)

const placements = Ahorn.PlacementDict(
	"Zipline (IsaGrabBag)" => Ahorn.EntityPlacement(
		Zipline,
		"rectangle"
	)
)

Ahorn.nodeLimits(entity::Zipline) = 0, 2

function Ahorn.selection(entity::Zipline)

	nodes = get(entity.data, "nodes", ())
    x, y = Ahorn.position(entity)

    res = [Ahorn.Rectangle(x - 8, y - 8, 16, 24)]
    
    for node in nodes
        nx, ny = Int.(node)
        
        newRes = Ahorn.Rectangle(nx - 8, y - 8, 16, 24)

        push!(res, newRes)
    end

    return res
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::Zipline, room::Maple.Room)
    
    ex, ey = Ahorn.position(entity)
	
	ey = ey - 1
	
	nodes = get(entity.data, "nodes", ())
	
	minx = ex
	maxx = ex
    
    for node in nodes
        nx, ny = Int.(node)
		
		if nx < minx
			minx = nx
		end
		if nx > maxx
			maxx = nx
		end
    end
	
	minx = minx - 10
	maxx = maxx + 10
	
	Ahorn.drawRectangle(ctx, minx - ex, -2, maxx - minx, 4, (0.0, 0.0, 0.0, 1.0), (0.0, 0.0, 0.0, 0.0))
	
	Ahorn.drawRectangle(ctx, minx - ex - 1, -4, 4, 8, (0.0, 0.0, 0.0, 1.0), (0.0, 0.0, 0.0, 0.0))
	Ahorn.drawRectangle(ctx, maxx - ex - 3, -4, 4, 8, (0.0, 0.0, 0.0, 1.0), (0.0, 0.0, 0.0, 0.0))
	
	
	Ahorn.drawRectangle(ctx, minx - ex, -1, maxx - minx, 1, (0.8, 0.84, 0.9, 1.0), (0.0, 0.0, 0.0, 0.0))
	Ahorn.drawRectangle(ctx, minx - ex, 0, maxx - minx, 1, (0.55, 0.5, 0.6, 1.0), (0.0, 0.0, 0.0, 0.0))
	
	Ahorn.drawRectangle(ctx, minx - ex - 0, -3, 2, 6, (0.8, 0.84, 0.9, 1.0), (0.0, 0.0, 0.0, 0.0))
	Ahorn.drawRectangle(ctx, maxx - ex - 2, -3, 2, 6, (0.8, 0.84, 0.9, 1.0), (0.0, 0.0, 0.0, 0.0))
    
    for node in nodes
        nx, ny = Int.(node)
		
		Ahorn.drawSprite(ctx, "isafriend/objects/zipline/handle_end.png", nx - ex, 4)
    end
	
	Ahorn.drawSprite(ctx, "isafriend/objects/zipline/handle.png", 0, 4)
end

end

File: Loenn/entities/zipline.lua
local drawableSprite = require("structs.drawable_sprite")
local drawableLine = require("structs.drawable_line")
local drawableRectangle = require("structs.drawable_rectangle")
local utils = require("utils")

local handleTex = "isafriend/objects/zipline/handle"
local handleEndTex = "isafriend/objects/zipline/handle_end"

local baseColor = {0.8, 0.84, 0.9, 1.0}
local highlightColor = {0.55, 0.5, 0.6, 1.0}

local width = 16
local halfWidth = 8

local zipline = {}

zipline.name = "isaBag/zipline"
zipline.depth = -500
zipline.nodeVisibility = "never"
zipline.nodeLimits = {0, 2}
zipline.placements = {
    name = "default",
    data = {
        usesStamina = true
    }
}

function zipline.sprite(room, entity)
    local sprites = {}
    
    local x, y = entity.x or 0, entity.y or 0
    local minX = x
    local maxX = x
    
    if entity.nodes then
        for _, node in ipairs(entity.nodes) do
            if node.x < minX then
                minX = node.x
            elseif node.x > maxX then
                maxX = node.x
            end
        end
    end
    
    -- Draw zipline + end markers
    local ziplineRect = utils.rectangle(minX - halfWidth, y - 6, maxX - minX + width, 4)
    local ziplineSprite = drawableRectangle.fromRectangle("bordered", ziplineRect, baseColor, {0, 0, 0})
    local points = {minX - halfWidth, y - 3.5, maxX + halfWidth, y - 3.5} -- Line has a half-pixel offset for some reason
    local highlightSprite = drawableLine.fromPoints(points, highlightColor, 1)
    table.insert(sprites, ziplineSprite)
    table.insert(sprites, highlightSprite)
    
    local leftEndRect = utils.rectangle(minX - halfWidth - 3, y - halfWidth, 4, 8)
    local rightEndRect = utils.rectangle(maxX + halfWidth - 1, y - halfWidth, 4, 8)
    local leftEndSprite = drawableRectangle.fromRectangle("bordered", leftEndRect, baseColor, {0, 0, 0})
    local rightEndSprite = drawableRectangle.fromRectangle("bordered", rightEndRect, baseColor, {0, 0, 0})
    table.insert(sprites, leftEndSprite)
    table.insert(sprites, rightEndSprite)
    
    -- Draw main handle
    local mainSprite = drawableSprite.fromTexture(handleTex, entity)
    table.insert(sprites, mainSprite)    
    
    -- Draw node handle(s)
    if entity.nodes then
        for _, node in ipairs(entity.nodes) do
            local nodeSprite = drawableSprite.fromTexture(handleEndTex, entity)
            nodeSprite:addPosition(node.x - x, 0)
            table.insert(sprites, nodeSprite)
        end
    end
    
    return sprites
end

function zipline.selection(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    
    local mainRect = utils.rectangle(x - halfWidth, y - halfWidth, 16, 17)
    local nodeRects = {}
    
    if entity.nodes then
        for _, node in ipairs(entity.nodes) do
            local nodeRect = utils.rectangle(node.x - halfWidth, y - halfWidth, 16, 17)    
            table.insert(nodeRects, nodeRect)
        end
    end
    
    return mainRect, nodeRects
end

-- Custom move function to limit node movement to the x-axis
function zipline.move(room, entity, node, offsetX, offsetY)
    local nodes = entity.nodes
    
    if node == 0 then
        entity.x += offsetX
        entity.y += offsetY
        
        -- When the main entity moves vertically, the zipline nodes should follow
        if nodes then
            for i, node in ipairs(entity.nodes) do
                if i <= #nodes then
                    node.y = entity.y
                end
            end
        end
        
        -- When a node moves, ignore any vertical movement
        else
            if nodes and node <= #nodes then
                local target = nodes[node]
                target.x += offsetX
            end
    end
end

-- The node selection should always match the height of the entity selection
function zipline.updateMoveSelection(room, entity, node, selection, offsetX, offsetY)
    if node == 0 then
        selection.x += offsetX
        selection.y += offsetY        
    else
        selection.x += offsetX
        selection.y = entity.y - halfWidth
    end
end

return zipline